### ISSUE TITLE
False positive: invalid-argument-type on valid pulumi Output.apply() lambda

### ISSUE THREAD

**Issue by yulius-balerion** [astral-sh/ty#2878]
## Description

`ty` reports `invalid-argument-type` on a valid `pulumi.Output.apply()` call using a lambda. The code is correct and works fine at runtime.

## Reproducer

```python
import pulumi

output: pulumi.Output[str] = pulumi.Output.from_input("hello")

result = output.apply(lambda s: s.upper())

pulumi.export("result", result)
```

## Reproduce Steps

1. Save file to `ty_fail.py`
2. Install `pulumi` however you want. (e.g. `brew install pulumi`)
3. Run `ty check ty_fail.py`

## ty output

```
warning[invalid-argument-type]: Argument to function `apply` is incorrect
 --> ty_demo.py:5:23
  |
3 | output: pulumi.Output[str] = pulumi.Output.from_input("hello")
4 |
5 | result = output.apply(lambda s: s.upper())
  |                       ^^^^^^^^^^^^^^^^^^^ Expected `Output[Literal["hello"]]`, found `(s) -> Unknown`
6 |
7 | pulumi.export("result", result)
  |
info: Function defined here
   --> .venv/lib/python3.12/site-packages/pulumi/output.py:222:9
    |
220 |         return self._is_secret
221 |
222 |     def apply(
    |         ^^^^^
223 |         self, func: Callable[[T_co], Input[U]], run_with_unknowns: bool = False
224 |     ) -> "Output[U]":
    |
info: Union variant `def apply[U](self, func: (Literal["hello"], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U]` is incompatible with this call site
```

## Root cause analysis

Two issues combine to produce the false positive:

1. **Over-narrowing**: `ty` infers the type of `pulumi.Output.from_input("hello")` as `Output[Literal["hello"]]` rather than `Output[str]`. The explicit annotation `output: pulumi.Output[str]` should prevent this narrowing but does not appear to constrain the overload resolution.

2. **Unresolvable overload**: The `.apply()` stubs in `pulumi/output.py` expose unbound overloads. When the inferred type is narrowed to `Literal["hello"]`, `ty` cannot match the lambda `(s) -> Unknown` against `(Literal["hello"], /) -> U | Awaitable[U] | Output[U]`.

The lambda `s: s.upper()` is perfectly valid — `str.upper()` is defined on `str`, and `Literal["hello"]` is a subtype of `str`.

## Environment

- `ty` version: ty 0.0.18 (751672710 2026-02-20)
- Python: Python 3.12.12
- Pulumi: v3.219.0

**Comment by sharkdp** [comment]
Thank you for reporting this. This should be fixed by #136.

**Comment by fritz-k** [comment]
@sharkdp I don't think the issue is caused by over-narrowing, since it persists even if you switched to a type that doesn't get narrowed to `Literal`.

Running `ty check` on
```python
import pulumi

output: pulumi.Output[dict[str, int]] = pulumi.Output.from_input({"i": 123})

result = output.apply(lambda d: d["i"] * 2)

pulumi.export("result", result)
```

produces
```bash
error[missing-argument]: No argument provided for required parameter `func` of function `apply`
 --> ty_fail.py:5:10
  |
3 | output: pulumi.Output[dict[str, int]] = pulumi.Output.from_input({"i": 123})
4 |
5 | result = output.apply(lambda d: d["i"] * 2)
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6 |
7 | pulumi.export("result", result)
  |
info: Union variant `def apply[U](self, func: (dict[str, int], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U]` is incompatible with this call site
info: Attempted to call union type `(bound method Output[dict[str, int]].apply[U](func: (dict[str, int], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U]) | (def apply[U](self, func: (dict[str, int], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U])`
info: rule `missing-argument` is enabled by default

error[invalid-argument-type]: Argument to function `apply` is incorrect
 --> ty_fail.py:5:23
  |
3 | output: pulumi.Output[dict[str, int]] = pulumi.Output.from_input({"i": 123})
4 |
5 | result = output.apply(lambda d: d["i"] * 2)
  |                       ^^^^^^^^^^^^^^^^^^^^ Expected `Output[dict[str, int]]`, found `(d) -> Unknown`
6 |
7 | pulumi.export("result", result)
  |
info: Function defined here
   --> /Users/kaisersi/dev/uapc/uapc/.direnv/python-3.12/lib/python3.12/site-packages/pulumi/output.py:222:9
    |
220 |         return self._is_secret
221 |
222 |     def apply(
    |         ^^^^^
223 |         self, func: Callable[[T_co], Input[U]], run_with_unknowns: bool = False
    |         ---- Parameter declared here
224 |     ) -> "Output[U]":
225 |         """
    |
info: Union variant `def apply[U](self, func: (dict[str, int], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U]` is incompatible with this call site
info: Attempted to call union type `(bound method Output[dict[str, int]].apply[U](func: (dict[str, int], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U]) | (def apply[U](self, func: (dict[str, int], /) -> U | Awaitable[U] | Output[U], run_with_unknowns: bool = False) -> Output[U])`
info: rule `invalid-argument-type` is enabled by default

Found 2 diagnostics
```

with
- ty `0.0.24 (876233049 2026-03-19)`
- Python `3.12.10`
- pulumi `3.226.0`

**Comment by benglewis** [comment]
@sharkdp This isn't fixed the latest version. Is there something that I should test?

**Comment by AlexWaygood** [comment]
> [@sharkdp](https://github.com/sharkdp) This isn't fixed the latest version. Is there something that I should test?

This was closed as a duplicate of #136. That issue is still open and has not yet been implemented.

**Comment by fritz-k** [comment]
@sharkdp I think my [comment above](https://github.com/astral-sh/ty/issues/2878#issuecomment-4116606373) got lost:

It looks like the issue isn't caused by over-narrowing, so #136 might not solve the issue here. If that's the case, you should probably reopen this issue here.

**Comment by sharkdp** [comment]
> [@sharkdp](https://github.com/sharkdp) I think my [comment above](https://github.com/astral-sh/ty/issues/2878#issuecomment-4116606373) got lost:

Yes, indeed. Sorry about that. I looked at your example again, and it looks like the problem is the same as in https://github.com/astral-sh/ty/issues/350. I added a [new comment](https://github.com/astral-sh/ty/issues/350#issuecomment-4235023260) in that issue to describe it further. If I comment out [this line](https://github.com/pulumi/pulumi/blob/b605630ec3db33517cd4297d41e2f69ca96c5cb8/sdk/python/lib/pulumi/output.py#L198) in pulumi's `Output` class, everything works fine.



> so [#136](https://github.com/astral-sh/ty/issues/136) might not solve the issue here

Ironically, solving #136 might actually solve #350 as well, since we would then rely on the declared type of `Output.apply`, which is set by the [function definition of `apply`](https://github.com/pulumi/pulumi/blob/b605630ec3db33517cd4297d41e2f69ca96c5cb8/sdk/python/lib/pulumi/output.py#L222-L224), and not affected by that implicit instance attribute assignment.
